### ISSUE TITLE
when did the configuration **default-database-strategy** work ?

### ISSUE THREAD

**Issue by jiangxiewei** [apache/incubator-shardingsphere#4464]
## First of all
my English is suck
## Question
when I only configured the default-database-strategy ( no specific table rule are configured ) , My defaul-database-strategy didn't work. (I use hintManager.setDatabaseShardingValue() )
##  Version
4.0.0
##  Detail
my configuration as follow :
``` properties
spring.shardingsphere.datasource.names=psghz,psgother
# psghz库配置
spring.shardingsphere.datasource.psghz.url=jdbc:postgresql://*******************:****/*****?useUnicode=true&characterEncoding=UTF-8&useSSL=false
spring.shardingsphere.datasource.psghz.username=****
spring.shardingsphere.datasource.psghz.password=********
spring.shardingsphere.datasource.psghz.driverClassName=org.postgresql.Driver
spring.shardingsphere.datasource.psghz.type=com.alibaba.druid.pool.DruidDataSource
spring.shardingsphere.datasource.psghz.max-active=10
# psgother库配置
spring.shardingsphere.datasource.psgother.url=jdbc:postgresql://*******************:****/*****?useUnicode=true&characterEncoding=UTF-8&useSSL=false
spring.shardingsphere.datasource.psgother.username=****
spring.shardingsphere.datasource.psgother.password=********
spring.shardingsphere.datasource.psgother.driverClassName=org.postgresql.Driver
spring.shardingsphere.datasource.psgother.type=com.alibaba.druid.pool.DruidDataSource
spring.shardingsphere.datasource.psgother.max-active=10
# 默认hint配置
spring.shardingsphere.sharding.default-database-strategy.hint.algorithm-class-name=com.ibuscloud.pubtrans.analysis.plugin.sharding.HintShardingStrategy
#起初打开default-data-srouce-name,结果走了DefaultDatabaseRoutingEngine,关闭则走了UnicastRoutingEngine
#spring.shardingsphere.sharding.default-data-source-name=psghz
spring.shardingsphere.props.sql.show=true
````

I follow  the code and see the RoutingEngineFactory.java create UnicastRoutingEngine.java 
``` java
         if (sqlStatementContext.getSqlStatement() instanceof DMLStatement && tableNames.isEmpty() && shardingRule.hasDefaultDataSourceName()) {
            return new DefaultDatabaseRoutingEngine(shardingRule, tableNames);
        }
        if (sqlStatementContext.getSqlStatement() instanceof DMLStatement && shardingConditions.isAlwaysFalse() || tableNames.isEmpty() || !shardingRule.tableRuleExists(tableNames)) {
            return new UnicastRoutingEngine(shardingRule, tableNames);
        }
```
and **!shardingRule.tableRuleExists(tableNames)** is true .
so back to the question , how does the configuration **default-database-strategy** work.

**Comment by jiangxiewei** [comment]
in 3.1.0 , 
when I use hintManager.setDatabaseShardingValue() , ShardingRouterFactory  will create DatabaseHintSQLRouter  as follow:
```java
    public static ShardingRouter newInstance(final ShardingRule shardingRule, final ShardingMetaData shardingMetaData, final DatabaseType databaseType, final boolean showSQL) {
        return HintManagerHolder.isDatabaseShardingOnly() ? new DatabaseHintSQLRouter(shardingRule, showSQL) : new ParsingSQLRouter(shardingRule, shardingMetaData, databaseType, showSQL);
    }
```

so I rollback to 3.1.0 and the default-database-strategy work well .

**Comment by wgy8283335** [comment]
OK. I will take care of this issue. You only concern configuration below right?
```
spring.shardingsphere.sharding.default-database-strategy.hint.algorithm-class-name=com.ibuscloud.pubtrans.analysis.plugin.sharding.HintShardingStrategy
```

**Comment by jiangxiewei** [comment]
> OK. I will take care of this issue. You only concern configuration below right?
> 
> ```
> spring.shardingsphere.sharding.default-database-strategy.hint.algorithm-class-name=com.ibuscloud.pubtrans.analysis.plugin.sharding.HintShardingStrategy
> ```
> 
> @jiangxiewei

yes... emmmm ,  I use hintManager.setDatabaseShardingValue() to add shardings in hint route to some certain sharding database without sharding tables And I only configured default-database-strategy.hint but it doesn't work like 3.1.0

**Comment by kimmking** [comment]
I think it's duplicated with https://github.com/apache/incubator-shardingsphere/issues/4478

**Comment by wgy8283335** [comment]
> > OK. I will take care of this issue. You only concern configuration below right?
> > ```
> > spring.shardingsphere.sharding.default-database-strategy.hint.algorithm-class-name=com.ibuscloud.pubtrans.analysis.plugin.sharding.HintShardingStrategy
> > ```
> > 
> > 
> > @jiangxiewei
> 
> yes... emmmm , I use hintManager.setDatabaseShardingValue() to add shardings in hint route to some certain sharding database without sharding tables And I only configured default-database-strategy.hint but it doesn't work like 3.1.0

Sharding-JDBC 4.0 is the software that you are using now, right?

**Comment by jiangxiewei** [comment]
> > > OK. I will take care of this issue. You only concern configuration below right?
> > > ```
> > > spring.shardingsphere.sharding.default-database-strategy.hint.algorithm-class-name=com.ibuscloud.pubtrans.analysis.plugin.sharding.HintShardingStrategy
> > > ```
> > > 
> > > 
> > > @jiangxiewei
> > 
> > 
> > yes... emmmm , I use hintManager.setDatabaseShardingValue() to add shardings in hint route to some certain sharding database without sharding tables And I only configured default-database-strategy.hint but it doesn't work like 3.1.0
> 
> Sharding-JDBC 4.0 is the software that you are using now, right? @jiangxiewei

no... I used 3.1.0 at first and I try to update the version 4.0.0

**Comment by jiangxiewei** [comment]
> > > OK. I will take care of this issue. You only concern configuration below right?
> > > ```
> > > spring.shardingsphere.sharding.default-database-strategy.hint.algorithm-class-name=com.ibuscloud.pubtrans.analysis.plugin.sharding.HintShardingStrategy
> > > ```
> > > 
> > > 
> > > @jiangxiewei
> > 
> > 
> > yes... emmmm , I use hintManager.setDatabaseShardingValue() to add shardings in hint route to some certain sharding database without sharding tables And I only configured default-database-strategy.hint but it doesn't work like 3.1.0
> 
> Sharding-JDBC 4.0 is the software that you are using now, right? @jiangxiewei

yes

**Comment by wgy8283335** [comment]
@jiangxiewei Hi, I find that a similar issue #3964. And the bug is fixed in 5.0.0 version.
